### PR TITLE
Move Codecov Ignores to LCOV

### PR DESCRIPTION
**Pull request by Ornias1993**
TL:DR
- Removes broken codecov ignore
- Places ignore section ax_code_coverage
- Forwards users from codecov to LCOV for ignores

Signed-off-by: Kjeld Schouten-Lebbing <kjeld@schouten-lebbing.nl>

### Motivation and Context
After many hours trying to figure out why we/I couldn't get codecov ignore working I have finally solved the problem.

The issue with codecov ignore came out to be two fold:
1. There is a bug in codecov, which prevents ignore from working, there have been about 3-5 reports about it from last months.
2. Codecov support is rather unsupportive for some reason, resorting to smokescreening and ignoring all reports of this bug.

This leads to my conclusion:
If we want to ignore folders, we beter not rely on codecov to do it for us.

### Description
This PR ignores folders, by removing them from the report before they  are send to codecov. As this is actually opensource and maintainable by OpenZFS/ZoL, I expect this to keep be a more sustainable long-term solution and this would even survive changing codecov-report-provider.

Before you ask: Why the wildcard prefix?
- The path thats actually ignored in the report, is the full path:
`Removing /var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/zfs-tests/cmd/file_check/file_check.c`
Not:
`tests/zfs-tests/cmd/file_check/file_check.c`

It's a bit of an oddity or bug with lcov, but it does not lead to sideeffects on the ZFS codebase.

Before you ask: Do we really need an ignore at all?
It would be cleaner to ignore tests covering themselves, but yes that wouldn't in itself warrant a change. However some PR's in the future might rely on externally maintained libraries that are not covered by our tests completely, an ignore would be essential to keep codecoverage reports representing actual coverage of zfs. (an example of which is/could be ZSTD)

### How Has This Been Tested?
Currently it's a draft. As soon as I get all tests to pass, i'll note it down and open it up for review.

### Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [X] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [X] I have updated the documentation accordingly.
- [X] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [X] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [X] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by Ornias1993** [comment]
@behlendorf , @freqlabs and @rlaager

I finally manage to get the problems with ignoring paths researched and got the ignore working. 💯

**Comment by freqlabs** [comment]
Interesting find! There is a much newer version of this file available, do you see anything we might get out of updating it?
http://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_code_coverage.m4

File: config/ax_code_coverage.m4 [hunk 1]
```diff
@@ -202,6 +202,10 @@ code-coverage-clean:
 # $(PACKAGE_VERSION). In order to add the current git hash to the title,
 # use the git-version-gen script, available online.
 
+# Add any folders you want to ignore here
+# Ignore tmp and tests themselves
+CODE_COVERAGE_IGNORE_PATTERN ?= "/tmp/*" "*/tests/*"
```
Comment (behlendorf) [inline config/ax_code_coverage.m4]: I'm curious why this was moved up?  Was it needed for some reason or purely cosmetic?

Comment (Ornias1993) [reply config/ax_code_coverage.m4]: Mainly cosmetic.

On row 209 it lists states:
`# Optional variables`

Most of which are not actually used by the current code.
It seems prudent to have non-optional variable that are actually used, before (and clearly seperated from), optional-unused variables.

However: If you prefer I could move the whole blob to row 222 and just use some extra spacing to make it stand out.

**Comment by Ornias1993** [comment]
@freqlabs 
It did cross my mind to look into it, but I limited myself to this fix only.

But: I don't think there is much gain in an update either, simply because it doesn't do that much for us and much of its features/flags do not seem to be used by us.

Besides: An update of this file, seems to be beter combined with a complete update of the autoconf stack and combining it with an added ignore flag seems out-of-scope of both.

File: config/ax_code_coverage.m4 (hunk 1 above)
Comment (Ornias1993) [reply config/ax_code_coverage.m4]: @behlendorf Moved it back down anyway.
In hindsight it looks just as clean... 👍

**Comment by Ornias1993** [comment]
Thanks for the fast feedback and review guys, appreciate it! :)

**Comment by rlaager** [comment]
.github/codecov.yml is missing the end-of-file newline. (You can see this in the GitHub diff viewer.)

**Comment by Ornias1993** [comment]
@rlaager  Thanks for noticing, totally missed that.
Fixed!
